mcp/streamable.go

```diff
@@ -797,56 +960,58 @@ func (s *streamableClientConn) Read(ctx context.Context) (jsonrpc.Message, error
 }
 
 // Write implements the [Connection] interface.
-func (s *streamableClientConn) Write(ctx context.Context, msg jsonrpc.Message) error {
-	s.mu.Lock()
-	if s.err != nil {
-		s.mu.Unlock()
-		return s.err
-	}
-
-	sessionID := s._sessionID
-	if sessionID == "" {
-		// Hold lock for the first request.
-		defer s.mu.Unlock()
-	} else {
-		s.mu.Unlock()
+func (c *streamableClientConn) Write(ctx context.Context, msg jsonrpc.Message) error {
+	data, err := jsonrpc2.EncodeMessage(msg)
+	if err != nil {
+		return err
 	}
 
-	gotSessionID, err := s.postMessage(ctx, sessionID, msg)
-	if err != nil {
-		if sessionID != "" {
-			// unlocked; lock to set err
-			s.mu.Lock()
-			defer s.mu.Unlock()
-		}
-		if s.err != nil {
-			s.err = err
-		}
+	if err := c.failure(); err != nil {
 		return err
 	}
 
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	sessionID := c._sessionID
 	if sessionID == "" {
-		// locked
-		s._sessionID = gotSessionID
+		sessionID, err := c.postMessage(ctx, "", data)
```

Comment:
This logic is a bit confusing:

It waits for a sessionID to be returned from postMessage but it could block in the case that handleSSE is called and is handling a stream.

It seems like we only _need_ to do this logic to wait for sessionID but the way it is written we are also waiting for the stream to complete.

Reply:
Should be a lot simpler now.